### ISSUE TITLE
Prover9 interface is sensitive to equation order

### ISSUE THREAD

**Issue by gebner** [gapt/gapt#318]
```scala
import at.logic.gapt.provers.prover9.Prover9Prover
import at.logic.gapt.formats.prover9.Prover9TermParserLadrStyle.parseFormula

new Prover9Prover().getLKProof(FSequent(
    Seq("(all x (all y (all z (x*(y*z) = (x*y)*z))))",
        "(all x (e*x = x))",
        "(all x (all y (i(x)*x = e)))")
      map parseFormula,
    Seq(parseFormula("a*i(a) = e"))))
```

This fails with:
```
at.logic.gapt.proofs.resolution.algorithms.ResolutionException: Resolution Exception: Clause [ :- =(*(*(v0, v1), v2), *(v0, *(v1, v2)))] is not reflexivity and not contained in CNF(-s) [
 :- =(*(x, *(y, z)), *(*(x, y), z));
 :- =(*(e, x), x);
 :- =(*(i(x), x), e);
=(*(a, i(a)), e) :- 
]
	at at.logic.gapt.proofs.resolution.algorithms.PCNF$.apply(PCNF.scala:69)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$PCNFMemoTable.getPCNF(robinson.scala:24)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$$anonfun$apply$1.apply(robinson.scala:36)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$$anonfun$apply$1.apply(robinson.scala:36)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:82)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:170)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:130)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:170)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:130)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:129)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:170)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:170)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:129)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:170)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:129)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:170)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.recConvert(robinson.scala:116)
	at at.logic.gapt.proofs.resolution.algorithms.RobinsonToLK$.apply(robinson.scala:36)
	at at.logic.gapt.provers.prover9.Prover9Prover.getLKProof(Prover9.scala:462)
	at Main$$anon$1.<init>(scratch.scala:10)
	at Main$.main(scratch.scala:1)
	at Main.main(scratch.scala)
```

(It works when I flip the equation in the associativity axiom.)

**Comment by frehn** [comment]
Thanks, that's definitely a bug since fixDerivation should flip the equation automatically. I'm having a look at that.

Regarding the title "Prover9 interface is sensitive to equation order", it may be that Prover9 is sensitive to equation order in general, so even though we can fix the bug described above, I think, although unlikely, we cannot rule out that prover9 may e.g. not find a proof anymore after flipping an equation.

**Comment by gebner** [comment]
> Thanks, that's definitely a bug since fixDerivation should flip the equation automatically. I'm having a look at that.

Thanks.

> I think, although unlikely, we cannot rule out that prover9 may e.g. not find a proof anymore after flipping an equation.

As far as I can tell, prover9 will reorder the equations using its internal ordering anyhow, so even if we changed the order before calling prover9, this shouldn't have any effect.

**Comment by frehn** [comment]
> As far as I can tell, prover9 will reorder the equations using its internal ordering anyhow, so even if we changed the order before calling prover9, this shouldn't have any effect.

That would certainly be reasonable - then I'd suggest to proceed by assuming that it is insensitive until a counterexample turns up :-)

**Comment by frehn** [comment]
Note to self towards resolving this bug: the problem in this case is that fixDerivation does not derive

:- =(*(*(v0, v1), v2), *(v0, *(v1, v2)))

from

:- =(*(x, *(y, z)), *(*(x, y), z))

via symmetry since it does not do variable renamings. fixDerivation would correctly derive

:- =(*(*(v0, v1), v2), *(v0, *(v1, v2)))

from

:- =(*(v0, *(v1, v2)), *(*(v0, v1), v2)).

The solution should be to incorporate variable renaming or even instantiation into the "deriveBySymmetry" code in fixDerivation.

**Comment by quicquid** [comment]
@frehn: it seems that we could use these strategies in different places within gapt. For instance, CERES currently just checks if a characteristic clause subsumes the additional content of the projection and applies substitutions and contractions accordingly. It might be helpful to have algorithms like fixSymmetry also available there. I think even a check if it can be fixed by symmetry might be useful.

**Comment by frehn** [comment]
Sounds good! The code in fixDerivation is quite modular - feel free to use it, or let me know if you need a different interface for this application.

**Comment by gebner** [comment]
This has been solved by merging #348.
